### ISSUE TITLE
Some nginx TLS tests started failing with LibreSSL 2.5.3 (but not with 2.4.4)

### ISSUE THREAD

**Issue by jirutka** [libressl-portable/portable#307]
After we updated LibreSSL from 2.4.4. to 2.5.3 in Alpine Linux, we have noticed that some TLS-related tests in nginx (both 1.10.3 and 1.12.0) started failing. Moreover, most of them fail because nginx accepted certificate that should be rejected! That’s pretty bad regression.

We’re not sure if the problem is in LibreSSL, nginx or nginx-tests, so reporting it to both. Nginx bug report: https://trac.nginx.org/nginx/ticket/1257.

People from VoidLinux have reproduced this issue too, on glibc.

```
Test Summary Report
-------------------
./h2_ssl_verify_client.t             (Wstat: 256 Tests: 5 Failed: 1)
  Failed test:  2
  Non-zero exit status: 1
./mail_imap_ssl.t                    (Wstat: 512 Tests: 14 Failed: 2)
  Failed tests:  4, 10
  Non-zero exit status: 2
./proxy_bind_transparent.t           (Wstat: 512 Tests: 3 Failed: 2)
  Failed tests:  1-2
  Non-zero exit status: 2
./proxy_ssl_certificate.t            (Wstat: 256 Tests: 7 Failed: 1)
  Failed test:  2
  Non-zero exit status: 1
./ssl_crl.t                          (Wstat: 512 Tests: 5 Failed: 2)
  Failed tests:  2-3
  Non-zero exit status: 2
./ssl_verify_client.t                (Wstat: 512 Tests: 12 Failed: 2)
  Failed tests:  4-5
  Non-zero exit status: 2
./ssl_verify_depth.t                 (Wstat: 256 Tests: 4 Failed: 1)
  Failed test:  2
  Non-zero exit status: 1
./stream_proxy_ssl_certificate.t     (Wstat: 256 Tests: 7 Failed: 1)
  Failed test:  2
  Non-zero exit status: 1
./stream_ssl_verify_client.t         (Wstat: 512 Tests: 12 Failed: 2)
  Failed tests:  3, 5
  Non-zero exit status: 2
Files=290, Tests=3628, 300 wallclock secs ( 1.91 usr  0.46 sys + 54.82 cusr  9.07 csys = 66.26 CPU)
```

Complete log: http://tpaste.us/Ynw6 or [nginx-tests.log.txt](https://github.com/libressl-portable/portable/files/955912/nginx-tests.log.txt)

/cc @ncopa

**Comment by Duncaen** [comment]
```
nginx-tests@openbsd$ TEST_NGINX_BINARY=/usr/local/sbin/nginx prove -v ./ssl_verify_client.t
./ssl_verify_client.t ..
1..12
ok 1 - plain connection
ok 2 - no cert
ok 3 - no optional cert
not ok 4 - bad optional cert

#   Failed test 'bad optional cert'
#   at ./ssl_verify_client.t line 125.
#                   'HTTP/1.1 200 OK
# Server: nginx/1.10.2
# Date: Tue, 25 Apr 2017 17:39:03 GMT
# Content-Type: text/plain
# Content-Length: 8
# Last-Modified: Tue, 25 Apr 2017 17:39:03 GMT
# Connection: close
# ETag: "58ff89b7-8"
# X-Verify: xSUCCESS:-----BEGIN CERTIFICATE-----
#       MIIBpzCCARACCQCgJ9TYR7fSOzANBgkqhkiG9w0BAQsFADAYMRYwFAYDVQQDDA0x
#       LmV4YW1wbGUuY29tMB4XDTE3MDQyNTE3MzkwM1oXDTE3MDUyNTE3MzkwM1owGDEW
#       MBQGA1UEAwwNMS5leGFtcGxlLmNvbTCBnzANBgkqhkiG9w0BAQEFAAOBjQAwgYkC
#       gYEAr70r4WM5yL5LJrp6rhBLOX6FKX85NY59H7J3/XHU1/wAnr/AUclq0ugfyj8e
#       rUTL5tNG8VC0fhKZ+nzgiMhY/WnLvF8LJArjw3DgeB8G5JAVZIMhUtfl9HmBv35Z
#       OhcFVgx315umUcDBSDlOAxES2wILhU7h6YB7qiQkHPYZN4kCAwEAATANBgkqhkiG
#       9w0BAQsFAAOBgQBkUOogRBmb+bANFb/o01+q4yC9o7mWmzL2YCgokJU6sw6PrY6Q
#       fExdPKrSYLwcQrGGow/zYMyndGRhVp0Z/61Ph6c7xXIVNge4L0JjVPZ4llgVcZTh
#       1FnvG808QBxPWY2s2lgmnu0B262DepxVHgDhR4oV8VUQRLMUMOnJFBBsWQ==
#       -----END CERTIFICATE-----x
# Accept-Ranges: bytes
#
# SEE-THIS'
#     doesn't match '(?^:400 Bad)'
not ok 5 - bad optional_no_ca cert

#   Failed test 'bad optional_no_ca cert'
#   at ./ssl_verify_client.t line 126.
#                   'HTTP/1.1 200 OK
# Server: nginx/1.10.2
# Date: Tue, 25 Apr 2017 17:39:03 GMT
# Content-Type: text/plain
# Content-Length: 8
# Last-Modified: Tue, 25 Apr 2017 17:39:03 GMT
# Connection: close
# ETag: "58ff89b7-8"
# X-Verify: xSUCCESS:-----BEGIN CERTIFICATE-----
#       MIIBpzCCARACCQCgJ9TYR7fSOzANBgkqhkiG9w0BAQsFADAYMRYwFAYDVQQDDA0x
#       LmV4YW1wbGUuY29tMB4XDTE3MDQyNTE3MzkwM1oXDTE3MDUyNTE3MzkwM1owGDEW
#       MBQGA1UEAwwNMS5leGFtcGxlLmNvbTCBnzANBgkqhkiG9w0BAQEFAAOBjQAwgYkC
#       gYEAr70r4WM5yL5LJrp6rhBLOX6FKX85NY59H7J3/XHU1/wAnr/AUclq0ugfyj8e
#       rUTL5tNG8VC0fhKZ+nzgiMhY/WnLvF8LJArjw3DgeB8G5JAVZIMhUtfl9HmBv35Z
#       OhcFVgx315umUcDBSDlOAxES2wILhU7h6YB7qiQkHPYZN4kCAwEAATANBgkqhkiG
#       9w0BAQsFAAOBgQBkUOogRBmb+bANFb/o01+q4yC9o7mWmzL2YCgokJU6sw6PrY6Q
#       fExdPKrSYLwcQrGGow/zYMyndGRhVp0Z/61Ph6c7xXIVNge4L0JjVPZ4llgVcZTh
#       1FnvG808QBxPWY2s2lgmnu0B262DepxVHgDhR4oV8VUQRLMUMOnJFBBsWQ==
#       -----END CERTIFICATE-----x
# Accept-Ranges: bytes
#
# SEE-THIS'
#     doesn't match '(?^:FAILED.*BEGIN)'
ok 6 - good cert
ok 7 - good cert optional
ok 8 - good cert trusted
ok 9 - no trusted sent
ok 10 - misdirected
ok 11 - no alerts
ok 12 - no sanitizer errors
# Looks like you failed 2 tests of 12.
Dubious, test returned 2 (wstat 512, 0x200)
Failed 2/12 subtests

Test Summary Report
-------------------
./ssl_verify_client.t (Wstat: 512 Tests: 12 Failed: 2)
  Failed tests:  4-5
  Non-zero exit status: 2
Files=1, Tests=12,  1 wallclock secs ( 0.04 usr  0.00 sys +  0.24 cusr  0.08 csys =  0.36 CPU)
Result: FAIL
```

Same or similiar results with the nginx port in `OpenBSD 6.1`.

**Comment by Duncaen** [comment]
I think its this commit [0] together with the nginxs `ngx_ssl_verify_callback` [1].

[0] https://github.com/libressl-portable/openbsd/commit/ddd98f8ea741a122952185a36c1396c14c2fda74 
[1] https://github.com/nginx/nginx/blob/branches/stable-1.10/src/event/ngx_event_openssl.c#L666

**Comment by jirutka** [comment]
I’ve tries to rebuild LibreSSL 2.5.3 with reverted https://github.com/libressl-portable/openbsd/commit/ddd98f8ea741a122952185a36c1396c14c2fda74 and now all nginx tests pass!

**Comment by jirutka** [comment]
Please read responses from the nginx team in https://trac.nginx.org/nginx/ticket/1257#comment:4. According to them it’s not problem on their said and it seems for us that they are right.

**Comment by 4a6f656c** [comment]
Thanks for digging into the issue and reporting this. It seems that the change does indeed break the documented API - we're looking into solutions and will follow up soon.

**Comment by kaniini** [comment]
this regression being a security issue, have we got a CVE yet?

**Comment by bob-beck** [comment]
I have seen no CVE for this issue. Perhaps nginx has one.

**Comment by bob-beck** [comment]
And before you ask, no we do not request them. If the discoverer of the
issue wishes to do so, that's fine.

**Comment by Shizmob** [comment]
Since this seems to affect at least two widely-used software packages severely if used together with LibreSSL, we opted to request a CVE for this.

**Comment by Shizmob** [comment]
CVE-2017-8301 has been assigned for this issue.

**Comment by 4a6f656c** [comment]
This has been reverted upstream - errata/release to follow soon.

**Comment by bob-beck** [comment]
https://ftp.openbsd.org/pub/OpenBSD/patches/6.1/common/003_libressl.patch.sig

**Comment by bob-beck** [comment]
Once the portable 2.5.4 heads out (shortly) I will close this out.

**Comment by bob-beck** [comment]
Errata releeased, and fixed in portable libressl 2.5.4

**Comment by kaniini** [comment]
thanks.
